### PR TITLE
[android] switch to clang & libc++

### PR DESCRIPTION
**Pull request by Rechi**
## Motivation and Context
Gcc is getting EOL in android ndk.

## How Has This Been Tested?
compiled depends and kodi for android with clang
(also compiled libffi for ios)

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: tools/depends/target/libffi/Makefile [hunk 1]
```diff
@@ -9,11 +9,6 @@ ARCHIVE=$(SOURCE).tar.gz
 
 # configuration settings
 CONFIGURE= ./configure --prefix=$(PREFIX) --disable-shared --disable-builddir 
-ifeq ($(OS), ios)
```
Comment (koying) [inline tools/depends/target/libffi/Makefile]: Why do you remove an ios patch?

**Review by koying — COMMENTED** [review]
Pretty limited, indeed :)
You might want to update the doc as to how to create a clang toolchain.

File: tools/depends/target/libffi/Makefile (hunk 1 above)
Comment (Rechi) [inline tools/depends/target/libffi/Makefile]: I had two options
- disable assembler for android
- fix the assembler code, which also fixes it for ios

I'll split the libffi changes into a separate PR.

Comment (koying) [inline tools/depends/target/libffi/Makefile]: Oh good. I feared it was some kind of accident...

**Comment by koying** [comment]
Made a build with libc++ on the latest NDK 16 beta and seems painless
Just have to package the actual lib, which is shared (by default ?)

**Comment by koying** [comment]
I have additional commits, but I can't seem to PR vs. your branch...
https://github.com/koying/xbmc/commits/androidClang

**Comment by Rechi** [comment]
@koying added your commits to this PR

**Comment by koying** [comment]
Thanks. Everything seems fine with clang + libc++
Here is the commit for libc++ switch: https://github.com/koying/xbmc/commit/2632c2e66dc675c8a5a312948015461aa78779b6
Basically, updating the doc and packaging libc++, which is shared by default.

I suggest to wait for NDK16 final, then ask for it to be installed with libc++, and go from there

**Comment by MartijnKaijser** [comment]
What is still missing for this?

**Comment by koying** [comment]
NDK16 final...Which came out yesterday or the day before ;)

**Comment by Paxxi** [comment]
Wohoo android no longer being the trouble child, awesome work!

**Comment by koying** [comment]
@Rechi mind picking https://github.com/koying/xbmc/commit/2632c2e66dc675c8a5a312948015461aa78779b6 (and maybe rename the PR to "switch to clang/libc++").

Thx

**Comment by koying** [comment]
jenkins build this please

**Comment by Rechi** [comment]
Not ready for merge yet, I have to fix the new compiler warnings first.

**Comment by MartijnKaijser** [comment]
@koying i tested this one on arm and arm64

**Review by koying — APPROVED** [review]

**Pull request merged**